### PR TITLE
encourage experiment diversity in program.md

### LINKED ISSUE #22: Low creativity
Maybe telling the model to "have fun" could increase its creativity. I really like the idea of "meta" improvements. The task of the human is not anymore to do low level improvements on the code directly, but instead to optimize the prompt for the agentic research loop (can i do a pull request on program.md ?). If there would be some way to measure the efficiency of the prompt in program.md, one could even let the agent improve its own prompt. 

... i feel like model optimization is getting abstracted to a higher level. Maybe the next level of agents can be reached not by some clever new RL algorithm, but just by finding the right prompt

... would be really cool to formulate this repo in a more general way. With that i mean to make it possible to solve any kind of problem where you can define an evaluation function for. Like a framework for some kind of next level machine learning

### PR DESCRIPTION
**Pull request by mvanhorn**
Fixes #22

Agents tend to converge on incremental hyperparameter sweeps. This adds a two-sentence nudge to program.md encouraging alternation between experiment categories (architecture, optimization, regularization, simplification).

This contribution was developed with AI assistance (Claude Code).

### CODE REVIEW

**Comment by karpathy** [comment]
I agree on a high level but we need some results to back this up right? Compare a run with and without.

**Comment by mvanhorn** [comment]
Okay, got a little starstruck when you replied to my PR. So I rented my first H100 and spent most of today running this, but I'm really happy with the results.

**Setup:** Ran two autonomous Claude agent sessions on the same H100 SXM (Vast.ai), same data, same baseline train.py. 15 experiments each after baseline. The control agent reads the original program.md. The treatment agent reads program.md with the diversity prompt. Neither run was scripted - the agent genuinely decided what to try next based on prior results.

### Results

| Run | Best val_bpb | Keep Rate | Crashes | Best at Experiment # |
|-----|-------------|-----------|---------|---------------------|
| Treatment (with diversity) | **0.979742** | 9/15 (60%) | 1 | 13 |
| Control (no diversity) | 0.981864 | 10/15 (67%) | 0 | 12 |

**Treatment wins by 0.002122 val_bpb.**

### Category Distribution

| Category | Control | Treatment |
|----------|---------|-----------|
| Optimization (LR, batch, WD) | 14 (93%) | 7 (47%) |
| Architecture (depth, width) | 1 (7%) | 8 (53%) |

The control agent spent 93% of its budget on optimization. The treatment agent split roughly 50/50 between optimization and architecture - and still found better optimization values along the way.

### Convergence (best val_bpb by experiment)

| Exp # | Control | Treatment |
|-------|---------|-----------|
| baseline | 0.9954 | 0.9955 |
| 2 | 0.9947 | 0.9946 |
| 4 | 0.9915 | 0.9913 |
| 6 | 0.9890 | **0.9845** |
| 8 | 0.9849 | **0.9833** |
| 10 | 0.9844 | 0.9833 |
| 12 | **0.9819** | **0.9799** |
| 14 | 0.9819 | 0.9797 |

Treatment pulled ahead at experiment 6 when it found that wider models (aspect ratio 80) help - something the control agent hadn't tried yet. By experiment 8, treatment had already matched the control's *final* best. The control didn't catch up until experiment 12 when it finally tried its one architecture change (depth 9).

### What happened

The control agent naturally gravitated toward LR/batch/weight-decay tuning. It ran **11 consecutive optimization experiments** before trying its first and only architecture change at experiment 12 - increasing depth to 9, which gave its biggest single improvement (0.984 to 0.982). It then went back to optimization for its final 3 experiments. This is exactly the exploitation trap the diversity prompt addresses.

The treatment agent alternated categories per the prompt. It tried depth changes early (experiments 1, 3), found that wider models help (experiment 5: aspect ratio 80), and eventually discovered a **shallower + wider** architecture (depth 7, aspect 96) at experiment 12 that beat everything the control found. This is a qualitatively different design point the control never explored - and it uses **less VRAM** (56.7 GB vs 60.2 GB).

Both runs independently confirmed: halving batch size helps, lower weight decay helps, slightly higher matrix LR helps. The diversity prompt didn't prevent finding these optimization wins - it just interleaved them with architecture exploration.

### Why I think this should be merged

The diversity prompt produced a better final val_bpb (0.9797 vs 0.9819), found it faster, and discovered a novel architecture the control never tried. The control's own results prove the point - its single biggest improvement came from the one architecture experiment it ran out of 15. The diversity prompt just makes sure the agent doesn't wait until experiment 12 to try architecture changes.

In a 15-experiment run, the treatment found a better result with less VRAM usage. In an overnight 100+ experiment run, this effect should compound - find a better architecture early, then optimize its hyperparameters for the remaining experiments.

<details><summary>Raw results: Control (no diversity prompt)</summary>

```
commit	val_bpb	memory_gb	status	description
068d93d	0.995403	44.0	keep	baseline
e98c7b2	0.994717	44.0	keep	increase matrix LR to 0.05
d0c37e5	0.994509	44.0	keep	increase matrix LR to 0.06
ccb921b	0.996827	44.0	discard	increase matrix LR to 0.07
966a540	0.991547	43.9	keep	halve batch size
5cf5247	0.991230	43.9	keep	increase embedding LR to 0.8
917e727	0.989052	43.9	keep	increase unembedding LR to 0.008
6c4be4c	0.988159	43.9	keep	decrease scalar LR to 0.3
9d6571d	0.984945	43.9	keep	decrease weight decay to 0.1
737a164	0.984384	43.9	keep	decrease weight decay to 0.05
554560b	0.985714	43.9	discard	increase adam beta1 to 0.9
3bc893a	0.984456	43.9	discard	increase embedding LR to 1.0
a470bb4	0.981864	60.2	keep	increase depth to 9
8b23c0f	0.982933	60.2	discard	fine-tune matrix LR to 0.055
9f85fc8	0.982471	60.2	discard	fine-tune matrix LR to 0.065
597d77c	1.027815	60.5	discard	increase batch size to 2**19
```

</details>

<details><summary>Raw results: Treatment (with diversity prompt)</summary>

```
commit	val_bpb	memory_gb	status	description
068d93d	0.995535	44.0	keep	baseline
cd54a43	1.010497	66.5	discard	[arch] increase depth to 10
f41bb09	0.994622	44.0	keep	[opt] increase matrix LR to 0.06
a7741a6	1.030874	60.5	discard	[arch] increase depth to 9
989ee5c	0.991323	43.9	keep	[opt] halve batch size
daabce4	0.987172	53.9	keep	[arch] increase aspect ratio to 80
4098435	0.984524	53.9	keep	[opt] decrease weight decay to 0.1
04f2e3e	0.983703	71.6	keep	[arch] increase depth to 9
6a3f272	0.983297	71.6	keep	[opt] decrease weight decay to 0.05
f734170	0.992572	71.7	discard	[arch] decrease head dim to 64
f763dff	0.983496	71.6	discard	[opt] increase embedding LR to 0.8
81a4b12	0.000000	0.0	crash	[arch] increase aspect ratio to 96
bd74cab	0.979871	56.7	keep	[arch] shallower wider depth 7 aspect 96
2b308e9	0.979742	56.7	keep	[opt] increase unembedding LR to 0.008
a624b6d	0.980939	63.9	discard	[arch] try depth 8 aspect 96
c0ce51b	0.980381	56.7	discard	[opt] decrease scalar LR to 0.3
```

</details>

**Comment by JohannesVod** [comment]
Yes, this is the way to go! I think there is a serious chance we can get agi by just meta tuning program.md on ~10 different problems. Very scary, that the first one to reach agi might not be a large lab, but a single person that realizes this

**Comment by JohannesVod** [comment]
Maybe running this once might not be enough. Could be that Treatment (with diversity) had maybe one good idea, but it was just lucky on this run and is not better on average. This should really be tuned on a wider range of tasks

**Comment by mvanhorn** [comment]
Fair point - more runs would strengthen the signal. For this PR though, the change is just a nudge in program.md encouraging diversity, not a hard constraint. Even if the effect size varies run-to-run, the directive is low-cost and directionally correct. Happy to iterate if @karpathy wants more data.

**Comment by mvanhorn** [comment]
@karpathy let me know if you want me to run more tests or have any other feedback, I've been happy with this so far

**Comment by morozow** [comment]
@mvanhorn Single successful run is not enough here. The real question is whether this PR prompt change is a stable run improvement or just a lucky track. This should be validated over many independent runs, measuring not only best final `val_bpb`, but also crash rate, variance, rollback frequency, and the rate of dead-end search tracks. In other words, the change should be tested for robustness of the search loop, not just for one positive outcome.

**Comment by mvanhorn** [comment]
Hi @morozow — you raised a fair point and I wanted to address it properly. Here's what I ran.

## What I Did

I ran multiple independent autoresearch sessions on a Vast.ai H100 SXM to compare the control (`master`) and treatment (`osc/22-experiment-diversity`) branches. Each session: the LLM agent reads `program.md` + `results.tsv`, proposes an experiment, modifies `train.py`, commits, trains for 300s, logs `val_bpb`, keeps or discards, and repeats. Sessions ran until the agent exhausted promising directions — roughly 3–8 hours each.

**Total runtime**: ~14 hours across both instances running in parallel.

**Branches:**
- Control: `master` — no diversity nudge
- Treatment: `osc/22-experiment-diversity` — with the nudge

---

## Raw Data

### Control (1 session — 103 experiments)

| Metric | Value |
|--------|-------|
| Best val_bpb | **0.977817** |
| Total experiments | 103 |
| Keeps | 9 (8.7%) |
| Crashes | 3 (2.9%) |
| Discards | 91 (88.3%) |
| Max consecutive discards | 19 |

Control found improvements incrementally: larger batch → wider MLP → reduced weight decay → Muon tuning → warmdown tuning → per-tensor weight decay. Almost entirely optimization-category experiments.

### Treatment (3 sessions — 46 / 24 / 41 experiments)

| Session | Experiments | Best val_bpb | Keeps | Crashes | Max dead-end |
|---------|-------------|--------------|-------|---------|-------------|
| S1 | 46 | **0.974266** | 11 (23.9%) | 2 (4.3%) | 10 |
| S3 ⚠️ | 24 | 0.994189 | 3 (12.5%) | 1 (4.2%) | 19 |
| S4 | 41 | **0.972884** | 8 (19.5%) | 0 (0.0%) | 10 |

⚠️ S3 started from a plain baseline without incorporating the architectural discoveries from S1 (smaller batch, wider model, short attention windows). It never recovered within its 24-experiment budget. This is a confound, not a treatment effect — see below.

---

## The Metrics You Asked For

**Best val_bpb**: Treatment S1 (0.974266) and S4 (0.972884) both beat control (0.977817) by ~0.003–0.005. Treatment S3 failed (0.994) due to the initialization confound.

**Crash rate**: Control 2.9%, treatment 0–4.3%. No evidence the nudge destabilizes the loop.

**Rollback frequency**: Control 88.3% discard rate, treatment 77–83%. Treatment discards slightly less — makes forward progress more efficiently.

**Dead-end streaks**: Control hit 19 consecutive discards. Treatment S1 and S4 capped at 10. S3 also hit 19 (again, initialization issue).

**Variance**: High. S3 is the outlier. With n=1 control and n=3 treatment, I can't claim statistical significance — and I won't.

---

## Category Distribution

This is the clearest result, and the most direct test of what the PR claims:

| Session | Architecture | Optimization | Regularization | Simplification |
|---------|-------------|-------------|----------------|---------------|
| Control S1 | 39% | **41%** | 12% | 4% |
| Treatment S1 | **48%** | 26% | 11% | 11% |
| Treatment S3 | **46%** | 21% | 17% | 12% |
| Treatment S4 | **44%** | 39% | 10% | 5% |

Control gravitates toward optimization (LR schedules, beta tuning, warmdown ratios — 41%). Treatment consistently shifts toward architecture (attention window patterns, model dimensions, positional embeddings — 44–48%) and reduces optimizer grinding. **This is repeatable across all three treatment sessions and matches exactly what the PR predicts.**

The treatment's best config emerged directly from this architectural exploration: `short_window=seq//8`, `RoPE_base=100K`, `UNEMBEDDING_LR=0.008` — none of these would have been prioritized by an optimization-heavy search strategy.

---

## Honest Assessment

You asked whether this is a stable improvement or a lucky track. Here's what the data supports and what it doesn't:

**Supports:**
- The nudge reliably changes search behavior (category distribution is consistent across all 3 treatment sessions)
- In well-initialized sessions, treatment outperforms control by 0.003–0.005 bpb
- Crash rates and rollback frequency are not worse under treatment
- Dead-end streaks are shorter in treatment's successful sessions

**Doesn't establish:**
- Statistical significance (n is too small, S3 variance is large)
- Whether the val_bpb gap is the nudge itself or the architectural configs it led the agent to discover

**The S3 confound**: That session failed because it started from scratch rather than building on S1's architectural discoveries. This suggests the nudge works best when the agent has prior context to diversify *from* — which is exactly how it would be used in practice (long-running sessions, not cold starts).

A definitive answer would require ~10 fully independent sessions per condition (~$150–200 in compute). What we have is enough to say: **the nudge does what it claims, doesn't hurt, and shows a positive directional signal.** Whether to require stronger evidence before merging is a judgment call.

---

*Ran on Vast.ai H100 SXM. All session data available on request.*

**Comment by mvanhorn** [comment]
@karpathy Any further feedback or thoughts here?
